### ISSUE TITLE
[Arc]...

### ISSUE THREAD

**Issue by pablo-CoCreator** [raycast/extensions#9758]
### Extension

https://raycast.com/the-browser-company/arc

#### Error:
```
SyntaxError: Expected ',' or '}' after property value in JSON at position 8729
    at JSON.parse (<anonymous>)
    at Object.E5 [as current] (/Users/thepmini/.config/raycast/extensions/606e26d8-b3c5-4760-8001-dccce1b872c4/search-tabs.js:87:20)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

### Steps To Reproduce

1. Try to use "Search Tabs" command of Arc Extension.
2. Error message displayed.

**Comment by pablo-CoCreator** [comment]
I'm not entirely sure if it's an access thing but Arc history seems to be working fine. Maybe make a "pairing" extension for Arc?

**Comment by tleo19** [comment]
Check https://github.com/raycast/extensions/issues/9109#issuecomment-1793713584
This is most likely a duplicate of #9109.

**Comment by megaroeny** [comment]
Been difficult to use this feature usually. Sometimes it works, usually it doesn't. Following and thanks!

**Comment by thomaslombart** [comment]
**This issue is now a bounty target** 💰

Hey there, we need help from our community to tackle this issue. **Working on it and having it fixed will allow you to get a month of Raycast Pro or a [mug](https://merch.raycast.com/product/mug-black).**

If you want to work on it, please let us know by posting a comment here with a rough estimate of how much time it'll take you to fix it. There are no hard deadlines, but we would like the issues to be addressed in a timely manner as this could be interesting for other contributors. You'll then be assigned the issue.

I'll be the point of contact for that. If you need more information, you can send me a message in the [Slack community](https://raycast.com/community), or in this issue. When creating your PR, please tag me either in the description or in a comment so that I can properly review it: @thomaslombart. 

I'll get in touch with you to give you the bounty once the issue is fixed.

Here's our guide about contributing to extensions: [Contribute to an extension](https://developers.raycast.com/basics/contribute-to-an-extension)

Looking forward to your contribution!

**Comment by thomaslombart** [comment]
Some pointers for this issue:

There's been a lot of issues opened on this extension that's related to JSON parsing. They seem to come in two forms:

- `SyntaxError: Unexpected token CHAR in JSON at position NUMBER`
- `SyntaxError: Bad escaped character in JSON at position NUMBER`

Most of the issues are closed because they got stale but it's still happening ([[5165](https://github.com/raycast/extensions/issues/5165)](https://github.com/raycast/extensions/issues/5165), [[5305](https://github.com/raycast/extensions/issues/5305)](https://github.com/raycast/extensions/issues/5305), [[6040](https://github.com/raycast/extensions/issues/6040)](https://github.com/raycast/extensions/issues/6040), [[7829](https://github.com/raycast/extensions/issues/7829)](https://github.com/raycast/extensions/issues/7829), [[8395](https://github.com/raycast/extensions/issues/8395)](https://github.com/raycast/extensions/issues/8395), [[8794](https://github.com/raycast/extensions/issues/8794)](https://github.com/raycast/extensions/issues/8794), [[9109](https://github.com/raycast/extensions/issues/9109)](https://github.com/raycast/extensions/issues/9109), [[9758](https://github.com/raycast/extensions/issues/9758)](https://github.com/raycast/extensions/issues/9758), [[9887](https://github.com/raycast/extensions/issues/9887)](https://github.com/raycast/extensions/issues/9887)). The root cause is probably due to a wrong JSON data that is produced through an AppleScript code. 

I've tested the following characters in an Arc tab and they all seem to work: . + * ? [ ^ ] $ ( ) { } = ! < > | : -

Only the backslash (`\`) isn't working and produces a `Bad escaped character` error. So, I'm not sure about all of the other errors.

Some potential solutions that could address this issue:
- Converting the AppleScript code to JXA but note that the API is undocumented and confusing
- Using a JSON library like [this one](https://github.com/mgax/applescript-json)
- Or, we could also use some [Foundation stuff](https://developer.apple.com/forums/thread/655669) to do this but it seems overly-complex

**Comment by loris** [comment]
Also not able to reproduce the issue except with the `\` character. I have renamed a tab `Ǵȑǻţįƈ ^$.?*'|+A-Zaz09_!~() ẗřȍŧŧįƞɇ 当你看到这个时` and it properly load in the Raycast extension, so i'm not sure neither about the other errors

**Comment by xilopaint** [comment]
Hey @thomaslombart, I got it to work with `\`. Example `foo \ bar`:

<img width="768" alt="Screenshot 2024-02-14 at 08 14 58" src="https://github.com/raycast/extensions/assets/11052366/2fa822d6-e1d6-4c07-b099-6a7827af8342">

I'll submit a PR.

**Comment by xilopaint** [comment]
Just checking all the related issues in this moment to make sure my solution covers all of them.

**Comment by xilopaint** [comment]
@thomaslombart Just submitted #10756 with the fix.

**Comment by xilopaint** [comment]
> Been difficult to use this feature usually. Sometimes it works, usually it doesn't. Following and thanks!

@megaroeny My PR fixes the issue related to the use of `\` in tab titles. If you encounter the issue with tabs that don't contain a backslash, please let me know.

**Comment by megaroeny** [comment]
> > Been difficult to use this feature usually. Sometimes it works, usually it doesn't. Following and thanks!
> 
> @megaroeny My PR fixes the issue related to the use of `\` in tab titles. If you encounter the issue with tabs that don't contain a backslash, please let me know.

Sounds good! I typically don't notice tab titles with a backslash. The last time I tried today, it would still open to a blank black tab. 🤷🏼‍♂️ 

How do I know if I received this fix/latest version? Is there a way to see a version number in Raycast?

**Comment by xilopaint** [comment]
> How do I know if I received this fix/latest version? Is there a way to see a version number in Raycast?

You can subscribe to notifications for #10756. The fix will be available once the PR is merged.

**Comment by pernielsentikaer** [comment]
> How do I know if I received this fix/latest version? Is there a way to see a version number in Raycast?

It's now merged 🙂

You can run the command `Check For Extension Updates`. Raycast also checks every fourth hour for new updates

**Comment by megaroeny** [comment]
> It's now merged 🙂
> 
> You can run the command `Check For Extension Updates`. Raycast also checks every fourth hour for new updates

Seems to work! Tried a few favorites across multiple Spaces, and it (eventually) goes to that tab! I'm sure the performance aspect is challenging, depending on how many tabs/favs/etc folks have. But it works! 👏

**Comment by megaroeny** [comment]
Okay, Arc _just_ got an update, and the extension/command keeps failing now. And then it prompts me to reload it. 😢 I'll attempt reinstalling it to see if it helps, but try updating Arc and test again. 🤦🏼

**Comment by xilopaint** [comment]
> Okay, Arc _just_ got an update, and the extension/command keeps failing now. And then it prompts me to reload it. 😢 I'll attempt reinstalling it to see if it helps, but try updating Arc and test again/monitor it. Sorry! 🤦🏼
> 
> It does work once or twice, but then breaks (on my end at least), since the `Version 1.30.0 (46409)` update.

@megaroeny It is unlikely that the issue is related to the Arc version. The extension works fine for me on v1.30.0. The most likely scenario is that it is related to the URL or title of the tabs you have open. Could you determine which tab is causing the problem? To do this, you can try removing your open tabs one by one and running the `Search Tabs` command in Raycast each time.

**Comment by megaroeny** [comment]
> @megaroeny It is unlikely that the issue is related to the Arc version. The extension works fine for me on v1.30.0. The most likely scenario is that it is related to the URL or title of the tabs you have open. Could you determine which tab is causing the problem? To do this, you can try removing your open tabs one by one and running the `Search Tabs` command in Raycast each time.

It seems to work now, yes. The last time I got the blank/black screen or the extension breaking, I believe it was a Twitch tab. But that doesn't have any `\` in it.

When I use it again tomorrow (ended my work day) I can follow-up more clearly.

**Comment by thomaslombart** [comment]
> Okay, Arc just got an update, and the extension/command keeps failing now

@megaroeny does it still fail with the same error? (`SyntaxError: Expected ... after property value in JSON at position ...`)

**Comment by megaroeny** [comment]
> @megaroeny does it still fail with the same error? (`SyntaxError: Expected ... after property value in JSON at position ...`)

Where would I see that? I haven't worked since Friday. When the extension crashed, it was a Raycast specific design/error message. Just mentioning to reload it.

I'll follow up this week when it happens again.

**Comment by xilopaint** [comment]
@megaroeny This doesn't look like a parsing issue. I suggest that you to open an issue with this content. Btw, do you have an old Mac and/or had an enormous amount of open tabs when the error was triggered? Do you see this error regardless of how many tabs you have open?

**Comment by xilopaint** [comment]
> I do have a bunch of bookmarks I've imported from my previous browser, so I don't lose those.

It doesn't make any difference if they are closed or not, the extension will parse them anyway. However, It would only be an issue if it is a huge number.

**Comment by megaroeny** [comment]
> Actually, It doesn't make any difference if they are closed or not, the extension will parse them anyway. However, It would only be relevant if it is a huge number.

I just did a quick check and it's probably around 1700 items... 🙈 That probably explains it. Maybe I should just get rid of that space and not use Arc in that manner... I'll give it a try and follow-up in #10847

**Comment by xilopaint** [comment]
> I just did a quick check and it's probably around 1700 items... 🙈 That probably explains it. Maybe I should just get rid of that space and not use Arc in that manner... I'll give it a try and follow-up in #10847

1700 items is a huge number and the likely cause of the error.

**Comment by xilopaint** [comment]
> Just deleted that huge space, and confirmed it was! Works _super snappy_ now! 🥳

AppleScript will always struggle to process large databases. Your issue could probably be solved by using `runAppleScript` from `@raycast/utils` instead of the `run-applescript` package and increasing the timeout. However, I'm not sure if we should address this kind of issue for an unreasonable amount of items, as it would also result in a slow UX that may not be well understood by the user.

**Comment by thomaslombart** [comment]
@xilopaint Are we good on the parsing issue, or is it still ongoing work? 🙂

**Comment by xilopaint** [comment]
@thomaslombart No ongoing work. If someone comes up with a new reproducible parsing issue we can fix it in a new PR. OP does not appear to be a regular GitHub user.

**Comment by megaroeny** [comment]
If anyone wants to do what I was doing and finds this issue, I have a solution as well, mentioned here: https://github.com/raycast/extensions/issues/10847#issuecomment-1959532140

Thx for the work on this y'all. 🙏🏼
